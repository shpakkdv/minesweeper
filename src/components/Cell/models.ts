import * as CellFieldModels from 'containers/Cell/models';
import * as GameFieldModels from 'containers/GameField/models';
import { Mines } from 'models';

export interface CellProps {
  x: number;
  y: number;
  value: number;
  mines: Mines;
  setMine: GameFieldModels.ActionCreator.SetMine;
  openItem: CellFieldModels.ActionCreator.OpenItem;
}
