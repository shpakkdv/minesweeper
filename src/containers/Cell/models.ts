import { IAction } from 'models';

export namespace Payload {
  export interface OpenItem {
    x: number;
    y: number;
  }
}

export namespace Action {
  export type OpenItem = IAction<Payload.OpenItem>;
}

export namespace ActionCreator {
  export type OpenItem = (x: number, y: number) => Action.OpenItem;
}
