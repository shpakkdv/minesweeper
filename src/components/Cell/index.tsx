import React from 'react';

import { CellProps } from './models';
import styles from './styles.module.scss';

const Cell: React.FunctionComponent<CellProps> = ({ x, y, value, mines, openItem, setMine }) => {
  const mine = mines[y][x];

  const onClick = () => {
    if (!mine) {
      openItem(x, y);
    }
  };

  const rightMouseClick: React.MouseEventHandler<HTMLButtonElement> = (event) => {
    event.preventDefault();
    setMine(x, y, !mine);
  };

  const opened = value !== -1;

  if (opened) {
    return (
      <div
        className={styles.OpenedCell}
        title={`x: ${x}, y: ${y}`}
      >
        {value === 0 ? '' : value}
      </div>
    );
  }

  return (
    <button
      className={styles.Cell}
      onContextMenu={rightMouseClick}
      onClick={onClick}
      title={`x: ${x}, y: ${y}`}
    >
      {mine ? '!' : ''}
    </button>
  );
};

export default Cell;
