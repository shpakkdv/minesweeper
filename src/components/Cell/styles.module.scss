@import 'styles/variables.scss';

@mixin cell {
  background-color: $cellColor;
  font-weight: bold;
  color: $fieldTableColor;
  user-select: none;
}

.Cell {
  @include cell;

  outline: $fieldLinesWidth solid $fieldTableColor;
  text-shadow: 0px 0px 10px #ff0000;
  cursor: pointer;
}

.OpenedCell {
  @include cell;

  display: flex;
  justify-content: center;
  align-items: center;
}
