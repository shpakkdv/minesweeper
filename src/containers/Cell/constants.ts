import { keyMirror } from 'utils/keyMirror';

export const ActionType = keyMirror(
  {
    OPEN_ITEM: null,
  },
  'Cell',
);
