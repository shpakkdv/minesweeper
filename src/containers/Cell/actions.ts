import { ActionType } from './constants';
import { ActionCreator } from './models';

export const openItem: ActionCreator.OpenItem = (x, y) => ({
  type: ActionType.OPEN_ITEM,
  payload: {
    x,
    y,
  },
});
