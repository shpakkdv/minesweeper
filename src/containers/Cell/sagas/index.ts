import { takeLatest, Effect } from 'redux-saga/effects';

import { ActionType } from '../constants';
import { openItem } from './openItem';

const effects: Effect[] = [
  takeLatest(ActionType.OPEN_ITEM, openItem),
];

export default effects;
