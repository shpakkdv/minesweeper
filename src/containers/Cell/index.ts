import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import Cell from 'components/Cell';
import { setMine } from 'containers/GameField/actions';
import { mines } from 'containers/GameField/selectors';
import { openItem } from './actions';

const mapStateToProps = createStructuredSelector({
  mines,
});

const mapDispatchToProps = {
  openItem,
  setMine,
};

export default connect(mapStateToProps, mapDispatchToProps)(Cell);
