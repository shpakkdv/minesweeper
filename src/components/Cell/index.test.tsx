import React from 'react'
import { shallow } from 'enzyme'

import Cell from './index';

describe('Cell', () => {
  it('Render a closed cell with a mine', () => {
    const wrapper = shallow(
      <Cell
        x={0}
        y={0}
        value={-1}
        mines={[[true]]}
        setMine={jest.fn()}
        openItem={jest.fn()}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('Render a closed cell without a mine', () => {
    const wrapper = shallow(
      <Cell
        x={0}
        y={0}
        value={-1}
        mines={[[false]]}
        setMine={jest.fn()}
        openItem={jest.fn()}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('Render an opened cell', () => {
    const wrapper = shallow(
      <Cell
        x={0}
        y={0}
        value={3}
        mines={[[false]]}
        setMine={jest.fn()}
        openItem={jest.fn()}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('Clicking on a closed cell without a mine should open it', () => {
    const openItem = jest.fn();
    const wrapper = shallow(
      <Cell
        x={0}
        y={0}
        value={-1}
        mines={[[false]]}
        setMine={jest.fn()}
        openItem={openItem}
      />
    );
    const button = wrapper.find('button');
    button.simulate('click');

    expect(button.text()).toBe('');
    expect(button.length).toBe(1);
    expect(openItem).toBeCalledTimes(1);
  });

  it('Clicking on closed cell with a mine should not open it', () => {
    const openItem = jest.fn();
    const wrapper = shallow(
      <Cell
        x={0}
        y={0}
        value={-1}
        mines={[[true]]}
        setMine={jest.fn()}
        openItem={openItem}
      />
    );
    const button = wrapper.find('button');
    button.simulate('click');

    expect(button.text()).toBe('!');
    expect(button.length).toBe(1);
    expect(openItem).toBeCalledTimes(0);
  });

  it('Invoking a context menu should set a mine', () => {
    const setMine = jest.fn();
    const wrapper = shallow(
      <Cell
        x={0}
        y={0}
        value={-1}
        mines={[[false]]}
        setMine={setMine}
        openItem={jest.fn()}
      />
    );
    const button = wrapper.find('button');
    button.simulate('contextmenu', { preventDefault: jest.fn() });

    expect(button.length).toBe(1);
    expect(button.text()).toBe('');
    expect(setMine).toBeCalledTimes(1);
  });

  it('An opened cell should not contain a button', () => {
    const wrapper = shallow(
      <Cell
        x={0}
        y={0}
        value={1}
        mines={[[false]]}
        setMine={jest.fn()}
        openItem={jest.fn()}
      />
    );
    const button = wrapper.find('button');

    expect(button.length).toBe(0);
  });
});
