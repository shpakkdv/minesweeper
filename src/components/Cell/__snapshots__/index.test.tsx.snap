// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Render a closed cell with a mine 1`] = `
<button
  className="Cell"
  onClick={[Function]}
  onContextMenu={[Function]}
  title="x: 0, y: 0"
>
  !
</button>
`;

exports[`Counter Render a closed cell without a mine 1`] = `
<button
  className="Cell"
  onClick={[Function]}
  onContextMenu={[Function]}
  title="x: 0, y: 0"
/>
`;

exports[`Counter Render an opened cell 1`] = `
<div
  className="OpenedCell"
  title="x: 0, y: 0"
>
  3
</div>
`;
